emma-language/src/main/scala/org/emmalanguage/compiler/lang/FunSupport.scala

```diff
@@ -0,0 +1,478 @@
+        frontier.headOption match {
+          case Some(v) => // vertex frontier is not empty
+            // collect lambdas defined and library functions refered in this vertex
+            val lambdas = lambdaDefsIn(v.tree)
+            val libdefs = libdefRefsIn(v.tree)
+
+            // collect the call edges in this vertex
+            val edgBldr = Set.newBuilder[CG.Edge]
+            // helper method: adds edges representing function parameter bindings
+            val addBnds = (argss: Seq[Seq[u.Tree]], sym: u.MethodSymbol) =>
+              for {
+                (args, params) <- argss zip sym.paramLists
+                (TermRef(arg), TermSym(param)) <- args zip params
+                if lambdas.contains(arg)
+              } edgBldr += CG.Binds(CG.FunPar(param), lambdas(arg))
+            // main traversal that accumulates the edges in `edgBldr`
+            TopDown
+              .inherit({
+                // inherit enclosing caller lambda or library function
+                case ValDef(sym, Lambda(_, _, _), _) =>
+                  lambdas.getOrElse(sym, v)
+              })(Monoids.right(v))
+              .inherit({
+                // inherit function type parameters of enclosing Lambda and DefDef nodes
+                case Lambda(_, params, _) =>
+                  params.map(Sym.of).filter(s => Sym.funs(Type.of(s).typeSymbol)).toList
+                case DefDef(_, _, _, paramss, _) =>
+                  paramss.flatten.map(Sym.of).filter(s => Sym.funs(Type.of(s).typeSymbol)).toList
+              })
+              .traverseWith({
+                // function parameter call
+                case Attr.inh(DefCall(Some(TermRef(tgt)), sym, _, _@_*), enclFunPars :: caller :: _)
+                  if enclFunPars.contains(tgt) =>
+                  edgBldr += CG.Calls(caller, CG.FunPar(tgt))
+                // lambda function call
+                case Attr.inh(DefCall(Some(TermRef(tgt)), sym, _, argss@_*), _ :: caller :: _)
+                  if lambdas.contains(tgt) =>
+                  edgBldr += CG.Calls(caller, lambdas(tgt))
+                  addBnds(argss, sym)
+                // library function call
+                case Attr.inh(DefCall(Some(TermRef(tgt)), sym, _, argss@_*), _ :: caller :: _)
+                  if LibDefRegistry(sym).isDefined =>
+                  edgBldr += CG.Calls(caller, CG.LibDef(sym, LibDefRegistry(sym).get))
+                  addBnds(argss, sym)
+              })(v.tree)
+            val edges = edgBldr.result()
+
+            // all parameters occurring in a `binds` edge
+            val funpars = edges.collect {
+              case CG.Binds(src, _) => src
+            }
+
+            // continue recursively
+            discover(
+              cg.copy(
+                vs = cg.vs ++ (lambdas.values ++ libdefs.values ++ funpars).toSet[CG.Vertex] + v,
+                es = cg.es ++ edges
+              ),
+              frontier.tail ++ (lambdas.values ++ libdefs.values).collect {
```

Comment:
The discovery `discover` methods iterates over a `frontier` of vertices with associated trees (that's why the type is `CG.TreeRef`). Since `CG.FunPar` vertices don't have one, we don't add them here (they won't match the `case` expression on line 326).